kotlinx-coroutines-core/common/src/sync/Semaphore.kt

```diff
@@ -123,74 +123,88 @@ private class SemaphoreImpl(private val permits: Int, acquiredPermits: Int) : Se
     override suspend fun acquire() {
         val p = _availablePermits.getAndDecrement()
         if (p > 0) return // permit acquired
-        addToQueueAndSuspend()
+        // While it looks better when the following function is inlined,
+        // it is important to make `suspend` function invocations in a way
+        // so that the tail-call optimization can be applied.
+        acquireSlowPath()
     }
 
-    override fun release() {
-        val p = incPermits()
-        if (p >= 0) return // no waiters
-        resumeNextFromQueue()
+    private suspend fun acquireSlowPath() {
```

Comment:
While this is a "slow path" it does not mean it should be inefficient. There is a way to improve a similar way to how channels are implemented. Please, call `suspendAtomicCancellableCoroutineReusable` in this function and pass the corresponding continuation to the `addToQueue` function (that would not be suspending anymore). This way, this Semaphore class wolud not allocate a state-machine object even in its slow-path.